vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1179 @@
+                    }
+                    if (vertex->links[i]->visited == vertex) {
+                        walker.emplace(vertex->links[i]);
+                    }
+                }
+            }
+            is_chanx = all_nulls ? all_nulls : is_chanx;
+            std::tuple<location, e_rr_type> key1(vertex->loc, e_rr_type::CHANY);
+            std::tuple<location, e_rr_type> key2(location(std::get<0>(vertex->loc), std::get<1>(vertex->loc) - 1), e_rr_type::CHANX);
+
+            if (is_chany) {
+                loc_chan_idx[key1] = chan_loc_map_[key1].size();
+                chan_loc_map_[key1].emplace_back(vertex->segment_id, R, C);
+            }
+            if (is_chanx) {
+                loc_chan_idx[key2] = chan_loc_map_[key2].size();
+                chan_loc_map_[key2].emplace_back(vertex->segment_id, R, C);
+            }
+            if (is_chanx && is_chany) {
+                shorts_.emplace_back(std::get<0>(key1), std::get<1>(key1), loc_chan_idx[key1],
+                                     std::get<0>(key2), std::get<1>(key2), loc_chan_idx[key2]);
+            }
+        }
+
+        /*
+         * Create edges between future rr_nodes that correspond to the given FPGA Interchange node
+         */
+        while (!back_walker.empty()) {
+            node_* vertex = back_walker.top();
+            back_walker.pop();
+            location loc = vertex->loc;
+            location down(std::get<0>(vertex->loc), std::get<1>(vertex->loc) - 1); // lower tile CHANY, our CHANX
+            location right(std::get<0>(down) + 1, std::get<1>(down));              // CHANX of tile to the right
+
+            if (loc_chan_idx.find(std::make_tuple(down, e_rr_type::CHANX)) != loc_chan_idx.end()) {
+                redirect_.emplace(std::make_tuple(node_id, loc), std::make_tuple(
+                                                                     down, e_rr_type::CHANX, loc_chan_idx[std::make_tuple(down, e_rr_type::CHANX)]));
+
+            } else if (loc_chan_idx.find(std::make_tuple(loc, e_rr_type::CHANY)) != loc_chan_idx.end()) {
+                redirect_.emplace(std::make_tuple(node_id, loc), std::make_tuple(
+                                                                     loc, e_rr_type::CHANY, loc_chan_idx[std::make_tuple(loc, e_rr_type::CHANY)]));
+
+            } else if (vertex->links[2] != nullptr && (vertex->links[2]->visited == vertex || vertex->links[2] == vertex->visited) && loc_chan_idx.find(std::make_tuple(right, e_rr_type::CHANX)) != loc_chan_idx.end()) {
+                redirect_.emplace(std::make_tuple(node_id, loc), std::make_tuple(
+                                                                     right, e_rr_type::CHANX, loc_chan_idx[std::make_tuple(right, e_rr_type::CHANX)]));
+
+            } else if (vertex->links[3] != nullptr && (vertex->links[3]->visited == vertex || vertex->links[3] == vertex->visited) && loc_chan_idx.find(std::make_tuple(down, e_rr_type::CHANY)) != loc_chan_idx.end()) {
```

Comment:
In general, it is better to split and precomupte a very long `if` condition to make it clearer.